Helper map header function

hayesla opened the pull request:
### Description
A first pass at creating helper functionality for creating a Map header from Issue #2278

At the moment I've put it in ```maputils.py``` as ```meta_helper``` class object. 
The object can be called through ```map.meta_helper``` . 


It currently print the accepted meta keywords that is used in ```GenericMap```  to create a map, and describes what each keyword means.

For example
```
>>> from sunpy import map
>>> map.meta_helper.acceptable_meta_keywords()
cunit1  :  Units of the coordinate increments along naxis1 e.g. arcsec **required
cunit2  :  Units of the coordinate increments along naxis2 e.g. arcsec **required
crval1  :  Coordinate value at reference point on naxis1 **required
crval2  :  Coordinate value at reference point on naxis2 **required
cdelt1  :  Spatial scale of pixels for naxis1, i.e. coordinate increment at reference point
cdelt2  :  Spatial scale of pixels for naxis2, i.e. coordinate increment at reference point
.....
```
It can also check if a meta keyword is accepted 
```
>>> map.meta_helper.acceptable_meta_keywords('crpix1')
crpix1  :  Pixel coordinate at reference point naxis1
>>> map.meta_helper.acceptable_meta_keywords('lalala')
lalala  is not in the accepted meta data required for sunpy.map.Map 
```
The class also has a utility function that can create a map header, stored in a `MetaDict`. This function can either take no keywords, and defaults to an Earth-based observer at current time (assuming cunit = arcsec), or can take a `SkyCoord` or `sunpy.coordinates.frame` as a reference point from which ctype, cunit, hglt/ln_obs, dsun_ref, rsun_ref and obstime can be pulled. It then can take optional crval, cdelt and crpix which are set to default 0, 1, center of data array if none are given.
  
For example

```
>>> from sunpy import map
>>> from sunpy.coordinates import frames
>>> from astropy.coordinates import SkyCoord
>>> from astropy import units as u
>>> import numpy as np

>>> data = np.random.rand(1024, 1024)
>>> header = map.meta_helper.make_header(data)
>>> header
MetaDict([('ctype1', 'HPLN-TAN'),
          ('ctype2', 'HPLT-TAN'),
          ('cunit1', 'arcsec'),
          ('cunit2', 'arcsec'),
          ('crpix1', 512.5),
          ('crpix2', 512.5),
          ('cdelt1', 1),
          ('cdelt2', 1),
          ('crval1', 0),
          ('crval2', 0),
          ('hgln_obs', 0.0),
          ('hglt_obs', 0.0),
          ('date_obs', '2019-04-30 19:29:47.129'),
          ('t_obs', '2019-04-30 19:29:47.129'),
          ('dsun_obs', 149597870700.0),
          ('rsun_obs', 959.2277285408243),
          ('rsun_ref', 695700000.0)])
```

or a `SkyCoord` or `frame` can be passed such that
```
>>> d = SkyCoord(0*u.arcsec, 0*u.arcsec, obstime="2017-08-01", observer = 'earth', frame=frames.Helioprojective)
>>> header = map.meta_helper.make_header(data, d)
>>> header
MetaDict([('ctype1', 'HPLN-TAN'),
          ('ctype2', 'HPLT-TAN'),
          ('cunit1', Unit("arcsec")),
          ('cunit2', Unit("arcsec")),
          ('crpix1', 512.5),
          ('crpix2', 512.5),
          ('cdelt1', 1),
          ('cdelt2', 1),
          ('crval1', 0),
          ('crval2', 0),
          ('hgln_obs', <Longitude 0. deg>),
          ('hglt_obs', <Latitude 5.78339799 deg>),
          ('date_obs',
           <Time object: scale='utc' format='isot' value=2017-08-01T00:00:00.000>),
          ('t_obs',
           <Time object: scale='utc' format='isot' value=2017-08-01T00:00:00.000>),
          ('dsun_obs', <Distance 1.51837236e+11 m>),
          ('rsun_obs', 945.0805956220959),
          ('rsun_ref', <Quantity 6.957e+08 m>)])
```
This is based on astropy.wcs.utils.celestial_frame_to_wcs. additonaly kwargs can be passed through `make_map` to define the metadict keywords

```
>>> header = map.meta_helper.make_header(data, d, crval = [100,100], cdelt = [10,10], exptime = 2, telescop = 'my_telescope', crota2 = -0.15)
```

I dont know if this is the type of functionality we want but it's a start I guess!

pep8speaks commented:
Hello @hayesla! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`sunpy/coordinates/wcs_utils.py`](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py):

> [Line 6:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L6): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 100 characters)
> [Line 56:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L56): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 56:49](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L56): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 56:51](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L56): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 61:8](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L61): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 62:20](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L62): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 62:22](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L62): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 63:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L63): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 68:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L68): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 71:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L71): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 73:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L73): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 94:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L94): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 95:28](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L95): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 96:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L96): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 100:74](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/coordinates/wcs_utils.py#L100): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`sunpy/map/maputils.py`](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py):

> [Line 25:84](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L25): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 26:82](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L26): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 29:85](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L29): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 32:88](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L32): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 42:57](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L42): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 45:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L45): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 100 characters)
> [Line 52:84](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L52): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 53:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L53): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 56:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 66:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L66): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 68:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L68): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 100 characters)
> [Line 85:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (156 > 100 characters)
> [Line 101:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 100 characters)
> [Line 108:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L108): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 100 characters)
> [Line 144:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L144): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 152:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L152): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 100 characters)
> [Line 157:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L157): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 261:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L261): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 264:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L264): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (152 > 100 characters)
> [Line 265:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L265): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (136 > 100 characters)
> [Line 269:1](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L269): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 271:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L271): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 272:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L272): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 272:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L272): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 273:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L273): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 274:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L274): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 275:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L275): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 275:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L275): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 100 characters)
> [Line 276:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L276): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 276:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L276): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 100 characters)
> [Line 277:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L277): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 278:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L278): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 279:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L279): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 280:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L280): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 281:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L281): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 282:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L282): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 283:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L283): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 284:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L284): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 285:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L285): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 286:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L286): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 287:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L287): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 288:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L288): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 289:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L289): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 290:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L290): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 291:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L291): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 292:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L292): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 293:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L293): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 294:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L294): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 295:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L295): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 296:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L296): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 297:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L297): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 298:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L298): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 299:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L299): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 300:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L300): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 301:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L301): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 302:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L302): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 302:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L302): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 303:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L303): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 303:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L303): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 304:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L304): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 304:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L304): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 305:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L305): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 305:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L305): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 306:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L306): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 306:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L306): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 307:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L307): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 307:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L307): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 308:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L308): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 308:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L308): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)
> [Line 309:19](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L309): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 309:101](https://github.com/sunpy/sunpy/blob/97fb6b3d66bf0571b6b668a7aad762a6e95dbf7b/sunpy/map/maputils.py#L309): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (131 > 100 characters)

wafels commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
 
-__all__ = []
 
+class map_helper:
```
CamelCase for classes, i.e, MapHelper

nabobalis commented on changelog/3083.feature.rst:
```diff
@@ -0,0 +1 @@
+Add a helper utility for creating a `sunpy.map.Map` header
```
How about:
```suggestion
Added a helper function that allows users to create a meta header for custom created `sunpy.map.Map`.
```

I think we want to expand on what its for and just to hopefully make it stand out in the changelog!

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
 
-__all__ = []
 
+class MapHelper:
```
Does this need to be a class and not just a function?

ehsteve commented:
@hayesla I think this is a great start! I would also recommend that this PR come with a guide (in addition to the normal documentation).

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+class MapHelper:
+    """
+    A helper class for knowing what is required in a `sunpy.map.Map` header.
+    A `sunpy.map.Map` header consists of a dictionary that contains meta information
```
A line break above here.

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    """
+    A helper class for knowing what is required in a `sunpy.map.Map` header.
+    A `sunpy.map.Map` header consists of a dictionary that contains meta information
+    about the map data. This meta is generally found in the FITS header of an image
```
```suggestion
    about the map data. This meta information is generally found in the FITS header of an image
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    A `sunpy.map.Map` header consists of a dictionary that contains meta information
+    about the map data. This meta is generally found in the FITS header of an image
+    file. This helper is defined to assist the user in finding out what meta data
+    keywords are accepted in a map header, and can produce a header MetaDict from a 
```
```suggestion
    keywords are accepted in a map header, and can produce a header `sunpy.util.meta.MetaDict from a 
```

I forgot where MetaDict is actually stored. That might not be correct in the slightest.

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    about the map data. This meta is generally found in the FITS header of an image
+    file. This helper is defined to assist the user in finding out what meta data
+    keywords are accepted in a map header, and can produce a header MetaDict from a 
+    coordinate object, or through default meta that can be then passed into `Map` 
```
```suggestion
    coordinate object, or through default meta that can be then passed into `sunpy.map.Map` 
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    file. This helper is defined to assist the user in finding out what meta data
+    keywords are accepted in a map header, and can produce a header MetaDict from a 
+    coordinate object, or through default meta that can be then passed into `Map` 
+    to create a Map object.
```
```suggestion
    to create a map.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    coordinate object, or through default meta that can be then passed into `Map` 
+    to create a Map object.
+
+    `map_helper.acceptable_meta_keywords()` prints the accepted meta that is used in 
```
```suggestion
    `sunpy.map.MapHelper.acceptable_meta_keywords()` prints the accepted meta that is used in 
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    to create a Map object.
+
+    `map_helper.acceptable_meta_keywords()` prints the accepted meta that is used in 
+    creating a sunpy.map.Map
```
```suggestion
    creating a `sunpy.map.Map`.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    `map_helper.acceptable_meta_keywords()` prints the accepted meta that is used in 
+    creating a sunpy.map.Map
+
+    `map_helper.make_header(data)` will create a map header for the numpy.ndarray data, 
```
```suggestion
    `sunpy.map.MapHelper.make_header(data)` will create a map header for the `numpy.ndarray` data, 
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    creating a sunpy.map.Map
+
+    `map_helper.make_header(data)` will create a map header for the numpy.ndarray data, 
+    either with default meta keywords, or a SkyCoord or frame can be passed to extract
```
```suggestion
    either with default meta keywords, or a `astropy.coordinates.SkyCoord` or frame can be passed to extract
```

I am not sure if that is right either. I forgot where SkyCoord lives.

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+    `map_helper.make_header(data)` will create a map header for the numpy.ndarray data, 
+    either with default meta keywords, or a SkyCoord or frame can be passed to extract
+    reference meta
```
```suggestion
    reference meta.
```

Does meta refer to "meta information" or to something else?

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    --------
+    Printing meta keywords:
+
+        * print the accepeted meta keywords that are used in a sunpy.map.Map header
```
```suggestion
        * Print the accepted meta keywords that are used in a `sunpy.map.Map` header:
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+          ...
+          CD2_2  :  Matrix element CDi_j describing the rotation required to align solar North with the top of the image.
+
+        * check is a meta keyword is acceptable and print information about it
```
```suggestion
        * Check is a meta keyword is acceptable and print information about it:
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    
+    Creating headers from map_helper:
+
+        * Create a default header with meta data assumed to be Earth-based observer at current time and arcsec units
```
```suggestion
        * Create a default header with meta data assumed to be Earth-based observer at current time and arcsec units:
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+          >>> header = map.MapHelper.make_header(data)
+          >>> my_map = map.Map(data, header)
+        
+        * Create a header from a reference SkyCoord object
```
```suggestion
        * Create a header from a reference SkyCoord object:
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+          >>> my_map = map.Map(data, header)
+
+
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+          >>> header = map.MapHelper.make_header(data, my_coord)
+          >>> my_map = map.Map(data, header)
+
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+          >>> my_coord = SkyCoord(100*u.arcsec, 100*u.arcsec, obstime = '2013-10-28 01:00', frame = frames.Helioprojective)
+          >>> header = map.MapHelper.make_header(data, my_coord)
+          >>> my_map = map.Map(data, header)
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    """
+    def acceptable_meta_keywords(self, *key):
+        """
+        prints the accepted metadata keywords that are used when
```
```suggestion
        Prints the accepted metadata keywords that are used when
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+    def make_header(self, data, *coordinate, crval=None, crpix=None, cdelt=None, **kwargs):
+        """
+        Function to create a MetaDict from a coordinate object (SkyCoord or coordinate frame)
```
```suggestion
        Function to create a `sunpy.util.meta.MetaDict` from a coordinate object (`astropy.coordinates.SkyCoord` or coordinate frame)
```

Have to check those paths first.

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        ----------
+        data : `~numpy.ndarray`
+            array data of Map for which a header is requried
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        coordinates : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`, optional if not set default parameters are set
+            coordinate object to get meta information for map header
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        crval : array or list of size of 2, default (0, 0)
+            coordinate of reference pixel along each axis
+            first index is crval1, second index is crval2
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        crpix : array or list of size of 2, default center of data array ((data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)
+            reference pixel along each axis
+            first index is crpix1, second index is crpix2
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        cdelt : array or list of size of 2, default (0, 0)
+            pixel scaling along each axis (i.e. the spatial scale of the pixels)
+            first index is cdelt1, second index is cdelt2
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        Parameters
+        ----------
+        data : `~numpy.ndarray`
+            array data of Map for which a header is requried
```
```suggestion
            Array data of Map for which a header is required.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        data : `~numpy.ndarray`
+            array data of Map for which a header is requried
+
+        coordinates : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`, optional if not set default parameters are set
```
```suggestion
        coordinates : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`, optional
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            array data of Map for which a header is requried
+
+        coordinates : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`, optional if not set default parameters are set
+            coordinate object to get meta information for map header
```
```suggestion
            Coordinate object to get meta information for map header.
             If not set, default parameters are used.
```

What are the default parameters?

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        coordinates : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`, optional if not set default parameters are set
+            coordinate object to get meta information for map header
+
+        crval : array or list of size of 2, default (0, 0)
```
```suggestion
        crval : `numpy.array` or `list` of size of 2, optional
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            coordinate object to get meta information for map header
+
+        crval : array or list of size of 2, default (0, 0)
+            coordinate of reference pixel along each axis
```
```suggestion
            Coordinate of reference pixel along each axis
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        crval : array or list of size of 2, default (0, 0)
+            coordinate of reference pixel along each axis
+            first index is crval1, second index is crval2
```
```suggestion
            first index is ``crval1``, second index is ``crval2``
            Defaults to ``(0, 0)``.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            first index is crval1, second index is crval2
+
+        crpix : array or list of size of 2, default center of data array ((data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)
+            reference pixel along each axis
```
```suggestion
            Reference pixel along each axis
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            coordinate of reference pixel along each axis
+            first index is crval1, second index is crval2
+
+        crpix : array or list of size of 2, default center of data array ((data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)
```
```suggestion
        crpix : `numpy.array` or `list` of size of 2, optional
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        crpix : array or list of size of 2, default center of data array ((data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)
+            reference pixel along each axis
+            first index is crpix1, second index is crpix2
```
```suggestion
            first index is ``crpix1``, second index is ``crpix2``.
            Defaults to the center of data array, ``((data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)``.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            first index is crpix1, second index is crpix2
+
+        cdelt : array or list of size of 2, default (0, 0)
+            pixel scaling along each axis (i.e. the spatial scale of the pixels)
```
```suggestion
            Pixel scaling along each axis (i.e., the spatial scale of the pixels)
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            reference pixel along each axis
+            first index is crpix1, second index is crpix2
+
+        cdelt : array or list of size of 2, default (0, 0)
```
```suggestion
        cdelt : `numpy.array` or `list` of size of 2, optional
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        cdelt : array or list of size of 2, default (0, 0)
+            pixel scaling along each axis (i.e. the spatial scale of the pixels)
+            first index is cdelt1, second index is cdelt2
```
```suggestion
            First index is ``cdelt1``, second index is ``cdelt2``.
            Defaults to ``(0, 0)``.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            pixel scaling along each axis (i.e. the spatial scale of the pixels)
+            first index is cdelt1, second index is cdelt2
+
+        **kwargs
```
```suggestion
        **kwargs: 
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        **kwargs
+            additional arguments that will be put into metadict header. These must be in the
+            list of acceptable meta keywords
```
```suggestion
            list of acceptable meta keywords.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            first index is cdelt1, second index is cdelt2
+
+        **kwargs
+            additional arguments that will be put into metadict header. These must be in the
```
```suggestion
            Additional arguments that will be put into metadict header. These must be in the
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        **kwargs
+            additional arguments that will be put into metadict header. These must be in the
+            list of acceptable meta keywords
+
```
I don't see where the `**kwargs`` are passed in. Maybe I'm blind.

hayesla replied:
in the last line in the function 
```for key in kwargs: ```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        Returns
+        -------
+        meta_dict : `~sunpy.util.MetaDict`
```
```suggestion
        `~sunpy.util.MetaDict`
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        Returns
+        -------
+        meta_dict : `~sunpy.util.MetaDict`
+            a MetaDict that contains the header information required for making a Map
```
```suggestion
            A `~sunpy.util.MetaDict` that contains the header information required for making a `sunpy.map.Map`.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        >>> my_coord = SkyCoord(0*u.arcsec, 0*u.arcsec, obstime="2017-08-01", observer = 'earth', frame=frames.Helioprojective)
+        >>> my_header = map.MapHelper.make_header(data, my_coord, crval = [0,0], cdelt = [2,2])
+        >>> my_map = map.Map(data, my_header)
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+        if coordinate:
+            if not isinstance(coordinate[0], (SkyCoord, frames.BaseCoordinateFrame)):
+                raise ValueError('coordinate needs to be an Astropy coordinate frame or SkyCoord instance.')
```
```suggestion
                raise ValueError('Input coordinate needs to be an Astropy coordinate frame or SkyCoord instance.')
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            meta = _get_meta_from_coordinate(coordinate[0])
+
+        else:
+            print('Using default meta assuming Earth based observer at current time with `arcsec` cunit type')
```
Should be the logger here?

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+                raise ValueError('coordinate needs to be an Astropy coordinate frame or SkyCoord instance.')
+
+            meta = _get_meta_from_coordinate(coordinate[0])
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        if coordinate:
+            if not isinstance(coordinate[0], (SkyCoord, frames.BaseCoordinateFrame)):
+                raise ValueError('coordinate needs to be an Astropy coordinate frame or SkyCoord instance.')
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+def _default_meta():
+    """
+    Creates default meta data if none is specified
```
```suggestion
    Creates a `dict` with the basic metadata.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+def _get_meta_from_coordinate(coordinate):
+    """
+    Function to get WCS meta data from a SkyCoord or frame using
+    `astropy.wcs.utils.celestial_frame_to_wcs`
```
```suggestion
    `astropy.wcs.utils.celestial_frame_to_wcs`.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+def _get_meta_from_coordinate(coordinate):
+    """
+    Function to get WCS meta data from a SkyCoord or frame using
```
```suggestion
    Function to get WCS meta data from a `~astropy.coordinates.SkyCoord` or frame using
```

Not sure if that is right.

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    Parameters
+    ----------
+    coordinate : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseFrame`
+        coordinate object to pull meta data from
```
```suggestion
        Coordinate object to pull metadata from.
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+    Returns
+    -------
+    coord_meta : dict
```
```suggestion
    `dict`
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    Returns
+    -------
+    coord_meta : dict
+        containing the meta information required for `sunpy.map.Map` header
```
```suggestion
        Containing the meta information required for `sunpy.map.Map` header:
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+            * dsun_obs
+            * rsun_obs
+            * t_obs, date_obs
+
```
```suggestion
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+    coord_meta = {}
+
+    if not isinstance(coordinate, (SkyCoord, frames.BaseCoordinateFrame)):
+        raise ValueError('takes an Astropy coordinate frame or SkyCoord instance.')
```
```suggestion
        raise ValueError('Input must be an Astropy coordinate frame or SkyCoord instance.')
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+        skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+    elif isinstance(coordinate, frames.SunPyBaseCoordinateFrame):
+        if coordinate.obstime is None:
+            raise ValueError('frame needs an observation time')
```
```suggestion
            raise ValueError('Frame needs an observation time.')
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+    if isinstance(coordinate, SkyCoord):
+        if coordinate.obstime is None or coordinate.frame is None:
+            raise ValueError('SkyCoord needs an observation time and a frame')
```
```suggestion
            raise ValueError('SkyCoord needs an observation time and a frame.')
```

nabobalis commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,309 @@
+
+    return coord_meta
+
+MapHelper = MapHelper()
```
What if we made the function call lower case?

Cadair commented:
I think that it would be best to move this into it's own file (should `maputils` be a directory?) and drop the class (using the file to group all the parts).

I will go through the exact header logic tomorrow morning probably.

hayesla commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,308 @@
+        for k in acceptable_meta:
+            print(k, " : ", acceptable_meta[k])
+
+    def make_header(self, data, *coordinate, crval=None, crpix=None, cdelt=None, **kwargs):
```
yeah I agree, I think we should probably get rid of the default map, and just have detailed docs with examples

hayesla commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,308 @@
+        else:
+            meta['cdelt1'], meta['cdelt2'] = cdelt[0], cdelt[1]
+
+        meta_dict = MetaDict({
```
yes.. yes it does

hayesla commented on sunpy/map/maputils.py:
```diff
@@ -1,7 +1,308 @@
+        if coordinate.obstime is None or coordinate.frame is None:
+            raise ValueError("SkyCoord needs an observation time and a frame.")
+        skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+    elif isinstance(coordinate, frames.SunPyBaseCoordinateFrame):
```
but should it not need a frame also?

Cadair replied:
I am pretty sure `SkyCoord` always had a frame.

hayesla commented:
I've updated this PR but I think we need to think about what else we want the helper function to do. At the moment I've updated it so that its a function that now lives in `sunpy.map.maphelper.py`

The two functions that live here are 

`meta_keywords` which prints the list of accepted fits keywords that are used to make a `sunpy.map.GenericMap`. 

The other one is `make_header` which takes an `astropy.coordinates.Skycoord`, from which the `cravl` meta is taken from the units given in the SkyCoord object, then the `date_obs` is taken from the `obstime` given in the SkyCoord, and the `cunit`, `ctype`,`hgln_obs`, `hglt_obs`, `dsun_obs` can all been taken from the SkyCoord.frame and defined observed in the SkyCoord.
The function also takes two optional inputs of astropy quantities of `reference_pixel` and `scale` from which `crpix` and `cdelt` are taken from. 

I guess the question is if we also want a function that doesn't include the passing of a SkyCoord - something similar to the SSW `make_map` https://hesperia.gsfc.nasa.gov/ssw/gen/idl/maps/make_map.pro

such that it would look something like:

```
def make_header(data, date_obs = parse_time('now'), xcen_coord = 0.0, ycen_coord = 0.0, dx = 1.0, dy = 1.0, xunit = 'arcsec', yunit = 'arcsec', coord_system = 'Helioprojective'):

	meta = {}
	meta['crval1'] = xcen_coord
	meta['crval2'] = ycen_coord
	meta['cdelt1'] = dx
	meta['cdelt2'] = dy
	meta['crpix1'] = (data.shape[0] + 1)/2.
	meta['crpix2'] = (data.shape[1] + 1)/2.
	meta['cunit1'] = xunit
	meta['cunit2'] = yunit
	meta['date_obs'] = date_obs 

	if coord_system.lower() == ('helioprojective' or 'heliographicstoneyhurst'):
		ctype1, ctype2 = 'HPLN-TAN', 'HPLT-TAN'
	elif coord_system.lower() == ('heliographiccarrington'):
		ctype1, ctype2 = 'CRLN-TAN', 'CRLT-TAN'
	elif coord_system.lower() == ('heliocentric'):
		ctype1, ctype2 = 'SOLX', 'SOLY'

	meta['ctype1'] = ctype1
	meta['ctype2'] = ctype2

	return MetaDict(meta)
```

hayesla commented:
will be working on this all day tomorrow with the sprints - so any suggestions on what this should look like let me know!

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,200 @@
+        raise ValueError("Input must be an Astropy coordinate frame or SkyCoord instance.")
+
+    if isinstance(coordinate, SkyCoord):
+        if coordinate.obstime is None or coordinate.frame is None:
```
I don't think it's possible to have a `SkyCoord` without a frame. I suggest you convert a `SkyCoord` to a frame here:
```python
if isinstance(coordinate, SkyCoord):
    coordinate = coordinate.frame
```

and then just implement all the logic after that for frames and not SkyCoord.

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,200 @@
+    cunit1, cunit2 = skycoord_wcs.wcs.cunit
+    coord_meta = dict(skycoord_wcs.to_header())
+    coord_meta['cunit1'], coord_meta['cunit2'] = cunit1, cunit2  # this is done as to_header transforms cunit to deg
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon, skycoord_wcs.heliographic_observer.lat
```
All four of these assignments here should not be quantities, so they should all be using `to_value`.

hayesla replied:
is there a reason they shouldn't be quantities?

Cadair replied:
My concern is consistency with values that come out of headers of FITS files, which will be floats and not quantities. I worry that it's implicitly assumed that they are not in GenericMap (especially `.save`)

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,200 @@
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon, skycoord_wcs.heliographic_observer.lat
+    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to(u.m)
+    coord_meta['rsun_ref'] = skycoord_wcs.rsun.to(u.m)
+    coord_meta['rsun_obs'] = ((skycoord_wcs.rsun.to(u.AU)/skycoord_wcs.heliographic_observer.radius.to(u.AU)).decompose() * u.radian).to(u.arcsec).value
```
Is this equation correct? Shouldn't this be an angular quantity?

Also you don't need to to `.to(u.AU)` on both sides of this, the `Quantity` magic will handle all the required unit conversion for you.

hayesla replied:
Good question - I'm actually not too sure?

hayesla commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,200 @@
+
+def meta_keywords(*key):
+    """
+    Prints the accepted metadata keywords that are used when
```
yes so accepted meta keywords here means essentially what keywords are actually taken from the meta and used in the generation of the map, of course more can be given they just wont necessarily be called in `GenericMap` (for example #1832). I think this function is helpful so that the user can see what the actual keywords are being used - but it may be better suited on the docs? Or if there was a way we could scrape from `mapbase.py` so that it gets updated

Cadair replied:
> I think this function is helpful so that the user can see what the actual keywords are being used

I can see the utility here, and I think we should keep it, I do think we need a more appropriate word than "accepted" though.

> Or if there was a way we could scrape from mapbase.py so that it gets updated.

Not easily enough that it's worth doing in this PR, I have an idea or two about how we could go about doing this if we wanted to. I don't think this set of data is likely to change very often so I am happy to keep it as it is.

If this is just for reference then maybe we should just be auto-generating some documentation rather than providing it as a function?

ehsteve commented:
Just to sow discord. I'm not sure I am on board with the functional approach...

Cadair commented:
:grin: That @ehsteve  is because you love yourself a class.

To my mind there is no state being stored anywhere here and therefore no need for a class at all. The main API here is one function to generate a map header.

nabobalis commented:
That CI fail looks like something I need to fix and is unrelated to the PR

ehsteve commented:
@Cadair,  @hayesla and I discussed the design and scope for this PR and we should all discuss to all get on the same page. We could do that on the phone call or earlier if you think there will already be lots to do on the call.

hayesla commented:
So with these commits several things have changes:

* tried to fix all the comments  - still unsure is the dsun_obs calculation is correct, but it seems to be right when playing around with calculating it 

* The function `accepted_meta` is now called `map_meta_keywords`

* The function `make_meta` is now called `make_fitswcs_header` following discussion yesterday

* if a coordinate system other than helioprojective is given, the coordinate is transformed to helioprojective 

*the `_get_meta_from_coordinate` function is now broken into two separate functions `_get_wcs_meta` and `_get_observer_meta`

* added a function `get_instrument_meta` that helps the user pass more logical keywords i.e. ``instrument`` vs ``instrume``. I don't know if we want to keep this, but following some discussion yesterday I threw it in

An example of using this helper is in this [gist](https://gist.github.com/hayesla/4cfc7846ff3d58b6660df95be0b88c76)

hayesla commented:
@Cadair  @ehsteve  let me know if you have any comments

wafels commented:
The calculation or rsun_obs is a small angle approximation. It should be replaced with the full (arctan I think) calculation to account for all situations.

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    reference_pixel :`~astropy.units.Quantity` of size 2, optional
+        Reference pixel along each axis. These are expected to be Cartestian ordered, i.e
+        the first index is the x axis, second index is the y axis.
+        Defaults to the center of data array, ``(data.shape[0] + 1)/2., (data.shape[1] + 1)/2.)``.
```
This is back-to-front

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    scale : `~astropy.units.Quantity` of size 2, optional
+        Pixel scaling along x and y axis (i.e. the spatial scale of the pixels (dx, dy)). These are
+        expected to be Cartestian ordered, i.e [dx, dy].
+        Defaults to ``([1., 1.] arcsec)``.
```
```suggestion
        Defaults to ``([1., 1.] arcsec/pixel)``.
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    Examples
+    --------
+    >>> from sunpy import map
```
`map` is a Python builtin:

```suggestion
    >>> import sunpy.map
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    >>> data = np.random.rand(1024, 1024)
+    >>> my_coord = SkyCoord(0*u.arcsec, 0*u.arcsec, obstime="2017-08-01", observer = 'earth', frame=frames.Helioprojective)
+    >>> my_header = map.make_fits_wcs(data, my_coord)
+    >>> my_map = map.Map(data, my_header)
```
```suggestion
    >>> my_map = sunpy.map.Map(data, my_header)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    >>> data = np.random.rand(1024, 1024)
+    >>> my_coord = SkyCoord(0*u.arcsec, 0*u.arcsec, obstime="2017-08-01", observer = 'earth', frame=frames.Helioprojective)
+    >>> my_header = map.make_fits_wcs(data, my_coord)
```
```suggestion
    >>> my_header = sunpy.map.make_fits_wcs(data, my_coord)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    if coordinate.obstime is None:
+        raise ValueError("The coordinate needs an observation time, `obstime`.")
+
+    coordinate = coordinate.transform_to(frames.Helioprojective)
```
```suggestion
```

We shouldn't do this, it's perfectly valid to construct a FITS-WCS header in other frames, we just need to only add the observer info where an observer is specified.

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    """
+    coord_meta = {}
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
```
Do we want this function to also do this? My understanding after our chat on Wed was this one would *only* do the observer stuff?

hayesla replied:
yeah so at the moment it just does the observer stuff right? It just pulls it from the heliographic_observer info in the frame - should we pull it rather from the skycoord observer keyword rather than the frame?

Cadair replied:
Yeah I would suggest you use the coordinate directly.

hayesla replied:
ping @Cadair

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    >>> my_map = map.Map(data, my_header)
+    """
+
+    if not isinstance(coordinate, SkyCoord):
```
Should we also support frame objects in this function?

hayesla replied:
yeah so I originally had that - but if we want observer info some frames do not have that? what do you think?

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    meta_wcs = _get_wcs_meta(coordinate)
+
+    meta_observer = _get_observer_meta(coordinate)
```
```suggestion
    if hasattr(coordinate, "observer"):
        meta_observer = _get_observer_meta(coordinate)
```

hayesla replied:
oh following on the previous comment - should we just have that if the observer information isn't given we just let map deal with that rather than requiring it here?

Cadair replied:
yes :)

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    if scale is None:
+        scale = u.Quantity([1.0*u.arcsec, 1.0*u.arcsec])
+
+    meta_wcs['crval1'], meta_wcs['crval2'] = coordinate.Tx.to_value(meta_wcs['cunit1']), coordinate.Ty.to_value(meta_wcs['cunit2'])
```
We need to do this in a frame agnostic way, which is hard. I suggest for the moment, like we do in map we don't allow HCC coordinates (and only spherical ones) and do what we do in `mapbase` to get lon / lat:
https://github.com/sunpy/sunpy/blob/master/sunpy/map/mapbase.py#L254

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    return coord_meta
+
+
+def _get_instrument_meta(**kwargs):
```
As in the call I am not sure I am 100% on board with this, but I am happy to have it. It does however need a couple of tweaks:

1) It needs to pop the values from `kwargs` or else we will end up raising an error later on in the header gen function. (see in line suggestion)
2) We need to document all of these kwargs in the docstring for `make_fitswcs_header`.

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    for key in kwargs:
+        if key in conversion:
+            coord[conversion[key]] = kwargs[key]
```
```suggestion
            coord[conversion[key]] = kwargs.pop(key)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(), skycoord_wcs.heliographic_observer.lat.to_value()
+    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to(u.m).to_value()
+    coord_meta['rsun_ref'] = skycoord_wcs.rsun.to(u.m).to_value()
```
```suggestion
    coord_meta['rsun_ref'] = skycoord_wcs.to_value(u.m)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(), skycoord_wcs.heliographic_observer.lat.to_value()
+    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to(u.m).to_value()
```
```suggestion
    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to_value(u.m)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(), skycoord_wcs.heliographic_observer.lat.to_value()
```
```suggestion
    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(u.deg), skycoord_wcs.heliographic_observer.lat.to_value(u.deg)
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,246 @@
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(), skycoord_wcs.heliographic_observer.lat.to_value()
+    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to(u.m).to_value()
+    coord_meta['rsun_ref'] = skycoord_wcs.rsun.to(u.m).to_value()
+    coord_meta['rsun_obs'] = ((skycoord_wcs.rsun/skycoord_wcs.heliographic_observer.radius).decompose() * u.radian).to(u.arcsec).to_value()
```
As @wafels pointed out this needs to use the full equation not the small angle approximation.

Cadair replied:
```suggestion
    coord_meta['rsun_obs'] = ((skycoord_wcs.rsun/skycoord_wcs.heliographic_observer.radius).decompose() * u.radian).to_value(u.arcsec)
```

Cadair commented:
@hayesla I have left a bunch more in line comments. I have also gone through and resolved all the comments I think are sorted or out of date, so the open comments on the PR are all things that need looking at (I think)

hayesla commented:
Ok so I think I've fixed the open comments - let me know what you think @Cadair 

and any other comments/docstrings etc, still need to do some tests!

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,253 @@
+    if isinstance(coordinate, SkyCoord):
+        coordinate = coordinate.frame
+
+    if coordinate.name == 'heliocentric':
```
I would you replace this with an `isinstance` check: (I think that's the correct frame name)
```suggestion
    if isinstance(coordinate, frames.Heliocentric):
```

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,253 @@
+        coordinate = coordinate.frame
+
+    if coordinate.name == 'heliocentric':
+        raise ValueError("SunPy Map does not support three dimensional data "
```
I would reword this error to be less map specific. Just something along the lines of "this function currently does not support heliocentric coordinates."

Cadair commented on sunpy/map/maphelper.py:
```diff
@@ -0,0 +1,253 @@
+        meta_wcs.update(meta_observer)
+
+    meta_instrument = _get_instrument_meta(**kwargs)
+
```
```suggestion
```
